server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -677,14 +677,30 @@ private static Map<String, Mapper> buildMergedMappers(
                     // replaces an existing one.
                     if (objectMergeContext.getMapperBuilderContext().getMergeReason() == MergeReason.INDEX_TEMPLATE) {
                         putMergedMapper(mergedMappers, mergeWithMapper);
-                    } else {
+                    } else if (isConflictingDynamicMapping(objectMergeContext, mergeWithMapper, mergeIntoMapper) == false) {
                         putMergedMapper(mergedMappers, mergeIntoMapper.merge(mergeWithMapper, objectMergeContext));
                     }
                 }
             }
             return Map.copyOf(mergedMappers);
         }
 
+        /*
+         * We're ignoring the field if a dynamic mapping update tries to define a conflicting field type.
+         * This is caused by another index request with a different value racing to update the mappings.
+         * After updating the mappings, the index request will be re-tried and sees the updated mappings for this field.
+         * The updated mappings will then be taken into account when parsing the document
+         * (for example by coercing the value, ignore_malformed values, or failing the index request due to a type conflict).
+         */
+        private static boolean isConflictingDynamicMapping(
+            MapperMergeContext objectMergeContext,
+            Mapper mergeWithMapper,
+            Mapper mergeIntoMapper
+        ) {
+            return objectMergeContext.getMapperBuilderContext().getMergeReason().isAutoUpdate()
+                && mergeIntoMapper.typeName().equals(mergeWithMapper.typeName()) == false;
```

Comment:
For dynamic mappings, it seems unlikely that the same type is used but with different immutable mapping parameters. That's because usually, there's one canonical mapping based on the type of the field.

There are some theoretical scenarios when using dynamic templates that could lead to such a situation. For example, when using a differet match_mapping_type but then mapping to the same field type with different immutable mapping parameters. This seems like an extreme edge case that I'm not sure is worth adding complexity for. We can still do that if it turns out to be an issue in practice.

Reply:
OK, I am good with that. I think the behavior will be to error out that the mapping is incompatible (or some exception like that). As long as we do not infinitely loop :slightly_smiling_face: ...

Reply:
Yes, this will lead to an exception (similar to the one detailed in https://github.com/elastic/elasticsearch/issues/114228) in the worst case rather than an infinite loop.

Given that, would you approve this implementation?